install.sh

```diff
@@ -131,42 +132,116 @@ check_deprecation() {
+
     # Variable transformations in global git config
     # Can be applied to all versions without any problem
     OLD_CONFIG_VALUE=$(git config --global githooks.autoupdate.updateCloneUrl)
     if [ -n "$OLD_CONFIG_VALUE" ]; then
-        git config --global githooks.cloneUrl "$OLD_CONFIG_VALUE"
+        git config --global githooks.cloneUrl "$OLD_CONFIG_VALUE" || LEGACY_TRANSFORM_FAILURES="true"
     fi
 
     OLD_CONFIG_VALUE=$(git config --global githooks.autoupdate.updateCloneBranch)
     if [ -n "$OLD_CONFIG_VALUE" ]; then
-        git config --global githooks.cloneBranch "$OLD_CONFIG_VALUE"
+        git config --global githooks.cloneBranch "$OLD_CONFIG_VALUE" || LEGACY_TRANSFORM_FAILURES="true"
     fi
 
     OLD_CONFIG_VALUE=$(git config --global githooks.previous.searchdir)
     if [ -n "$OLD_CONFIG_VALUE" ]; then
-        git config --global githooks.previousSearchDir "$OLD_CONFIG_VALUE"
+        git config --global githooks.previousSearchDir "$OLD_CONFIG_VALUE" || LEGACY_TRANSFORM_FAILURES="true"
     fi
 
     # Copy legacy file to new location
     if [ -f "$INSTALL_DIR/autoupdate/registered" ]; then
-        cp "$INSTALL_DIR/autoupdate/registered" "$INSTALL_DIR/registered"
+        cp "$INSTALL_DIR/autoupdate/registered" "$INSTALL_DIR/registered" || LEGACY_TRANSFORM_FAILURES="true"
+    fi
+
+    # Check if we need to move .shared files entries to `--local githooks.shared`
+    # and split comma-separated `--global githooks.shared`
+
+    if [ "$LEGACY_TRANSFORM_FAILURES" = "true" ]; then
+        echo "! There were legacy transform errors: check stderr"
+        return 1
     fi
 
     return 0
 }
 
+############################################################
+# Tests if the commit before a feature is introduced `$2`
+#   is after or equal the commit before the update `$1`.
+#   Note: Any feature that this install.sh knows about
+#   is an ancestor of the current version.
+#
+# Returns: 0 if succesful, 1 otherwise
+############################################################
+legacy_transformation_update_includes() {
+    LAST_COMMIT="$1"    # Last commit of before the update.
+    FEATURE_BEFORE="$2" # The commit right before the feature is introduced.
+
+    if [ -n "$LAST_COMMIT" ] &&
+        execute_git "$GITHOOKS_CLONE_DIR" merge-base --is-ancestor \
+            "$LAST_COMMIT" "$FEATURE_BEFORE" >/dev/null 2>&1; then
+        # feature-before >= last commit
+        return 0
+    fi
+
+    return 1
+}
+
+############################################################
+# Function to dispatch to all legacy transformations
+#   right after the update.
+#
+# Returns: 0 if succesful, 1 otherwise
+############################################################
+legacy_transformations_after_update() {
+    COMMIT_COUNT=$(execute_git "$GITHOOKS_CLONE_DIR" rev-list --count HEAD)
+    CURRENT_COMMIT=$(execute_git "$GITHOOKS_CLONE_DIR" rev-parse HEAD)
+
+    if [ "$COMMIT_COUNT" != "1" ] && [ -z "$INTERNAL_UPDATED_FROM_COMMIT" ]; then
```

Comment:
What is the point of this COMMIT_COUNT check, btw? Is it just to check that we're actually in a repo? Or that it is not a shallow copy? Can the count ever be 1?

Reply:
I think: when `INTERNAL_UPDATED_FROM_COMMIT` is not set (old install, even with git release clone already)
we need to make sure its an update. `--internal-autoupdate` is set from `base-template.sh` but not from `cli.sh` (which is correct).
So when executing `git hooks update` with an old cli.sh or a simple call to `install.sh` (which is a fresh install), I think we need to make sure we have more than one commit, since a fresh install always clones with `--depth 1`

Reply:
Note: That `update_release_clone` is executed before the dispatch to the new `install.sh` so we have no notion if it was a new clone or an update. With this PR, we have not this correct set `INTERNAL_UPDATED_FROM_COMMIT` if it was an **update**.

Reply:
But this code (legacy_transformations_after_update) is only called *after* the update. In particular, it is called `    if ! is_postupdate; then`, which checks for `$INTERNAL_POSTUPDATE`. And since `INTERNAL_POSTUPDATE` is only set by `parse_command_line_arguments`, if `--internal-postupdate` is given. So:
 - Setting the default value of `INTERNAL_UPDATED_FROM_COMMIT` could move to the end of `update_release_clone` (inside an `if [ $INTERNAL_POSTUPDATE = "true"]`) and it would be applied in exactly the same cases as it is now.
 - When this code runs, `--internal-postupdate` is given, so (I think) it always runs post-upgrade, so this `--count` check is not really needed?
 - On a fresh install, this code does not run (since `--internal-postupdate` is not given), so that's not a case to be handled by this code.

There is the question of upgrades from very old versions, which I suspect do not pass `--internal-postupdate`? However, if there is no way to distinguish such an update from a fresh or manual install, then I don't think we can do any meaningful migration in that case?

And there might be a case of a manual upgrade using `install.sh`, but I think that updates and then just uses `--internal-postupdate`, so that's handled properly arleady?

Reply:
Huch, ok, I am not sure if I get your points correctly:
The thing is what ever we do inside the `if ! is_postupdate; then`  namely:
```shell
 if ! is_postupdate; then

        check_deprecation || return 1

        legacy_transformations_before_update || return 1

        update_release_clone || return 1

        if is_clone_updated || ! is_running_internal_install; then
            # Either
            # - we just updated the release clone -> dispatch to the new install
            # - or we are running not from the release clone -> dispatch to the install
            # in the clone.
            run_internal_install --internal-postupdate "$@" || return 1
            return 0
        fi
    fi
```

we directly loose as we 
do
```
run_internal_install --internal-postupdate "$@" || return 1
```

But you are probably right in a way like the last commit change, please have a look:
Moving this function there, is the right spot, and thats the only thing which can change IMO. I also renamed it to make it more clear (?).

Reply:
In 20ddf0c reduced the logic to what is needed, and is probably the way you think it should be, IMO this works. I need to test it yet...

Reply:
Hm, seems my analysis was not entirely correct, I assumed that `if ! post_update`, then script would always `return`, but in your quote you show that this is only `if is_clone_updated || ! is_running_internal_install; then`. I don't really see how all this interacts right now and can't seem to find the concentration and time to dive in, so if you indeed think things are ok now, then with your last commit, I think they are really a lot cleaner and nicer, so I think that's good enough :-)